File: experiments/go/url_checker/checker.go
package main

import (
     "net/http"
     "fmt"
)

// [{http://www.tumblr.com/audio_file/someonesinthewolf/8755310631/tumblr_lpkxbfQrNr1qbpu1l false}
// {http://www.tumblr.com/audio_file/someonesinthewolf/8755310631/tumblr_lpkxbfQrNr1qbpu1l?plead=please-dont-download-this-or-our-lawyers-wont-let-us-host-audio true}]
// real    0m0.139s
// user    0m0.011s
// sys 0m0.008s

type UrlMap struct {
    Url string
    Exists bool
}

func loadUrls() []string {
    // file_in, _ := ioutil.ReadFile("urls.txt")
    // txt := string(file_in)
    // return strings.Split(strings.TrimSpace(txt), "\n")
    return []string {"http://www.tumblr.com/audio_file/someonesinthewolf/8755310631/tumblr_lpkxbfQrNr1qbpu1l?plead=please-dont-download-this-or-our-lawyers-wont-let-us-host-audio", "http://www.tumblr.com/audio_file/someonesinthewolf/8755310631/tumblr_lpkxbfQrNr1qbpu1l"}
}

func check(returnChannel chan UrlMap, url string) {
    resp, _ := http.Head(url)
    exists := resp.StatusCode == 200

    returnChannel <- UrlMap{url, exists}
}

func wait(responseChannel chan UrlMap, numberOfUrls int) (mapping []UrlMap) {
   done := 0

   for {
        mapping = append(mapping, <- responseChannel)
        done++

        if done >= numberOfUrls {
            break
        }
    }
    return
}

func main() {
    urls := loadUrls()

    responseChannel := make(chan UrlMap)

    for _, url := range urls {
        go check(responseChannel, url)
    }

    mapping := wait(responseChannel, len(urls))

    fmt.Println(mapping)
}

File: experiments/go/url_checker/checker.js
var http = require('http'),
    parseUrl = require('url').parse;


// Lucass-MacBook-Air:go lucas$ time node checker.js
// [ { url: 'http://www.tumblr.com/audio_file/someonesinthewolf/8755310631/tumblr_lpkxbfQrNr1qbpu1l',
//     exists: false },
//   { url: 'http://www.tumblr.com/audio_file/someonesinthewolf/8755310631/tumblr_lpkxbfQrNr1qbpu1l?plead=please-dont-download-this-or-our-lawyers-wont-let-us-host-audio',
//     exists: false } ]

// real    0m0.185s
// user    0m0.085s
// sys 0m0.020s

function loadUrls(cb){
    var urls = ["http://www.tumblr.com/audio_file/someonesinthewolf/8755310631/tumblr_lpkxbfQrNr1qbpu1l?plead=please-dont-download-this-or-our-lawyers-wont-let-us-host-audio", "http://www.tumblr.com/audio_file/someonesinthewolf/8755310631/tumblr_lpkxbfQrNr1qbpu1l"];
    cb(urls);
}

function check(url, cb){
    var info = parseUrl(url),
            opts = {
            'method': 'HEAD',
            'port': 80,
            'host': info.host,
            'path': info.path
        };
    http.request(opts, function(res){
        cb(url, (res.statusCode === 200));
    }).end();
}

var res = [],
    remaining;

function done(){
    console.log(res);
}

loadUrls(function(urls){
    remaining = urls.length;
    urls.forEach(function(url){
        check(url, function(url, exists){
            res.push({'url': url, 'exists': exists});
            remaining--;
            if(remaining === 0){
                done();
            }
        });
    });
});